framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -566,170 +556,100 @@ protected void reportPurityErrors(
+    protected void checkContractsAtMethodDeclaration(
+            MethodTree node,
+            ExecutableElement methodElement,
+            List<String> formalParamNames,
+            boolean abstractMethod) {
         FlowExpressionContext flowExprContext = null;
-        Set<PreOrPostcondition> postconditions = contractsUtils.getPostconditions(methodElement);
+        List<Contract> contracts = contractsUtils.getContracts(methodElement);
 
-        for (PreOrPostcondition p : postconditions) {
-            String expression = p.expression;
-            AnnotationMirror annotation = AnnotationUtils.fromName(elements, p.annotationString);
+        for (Contract contract : contracts) {
+            String expression = contract.expression;
+            AnnotationMirror annotation = contract.annotation;
 
-            // Only check if the postcondition concerns this checker
-            if (!atypeFactory.isSupportedQualifier(annotation)) {
-                continue;
-            }
             if (flowExprContext == null) {
                 flowExprContext =
                         FlowExpressionContext.buildContextForMethodDeclaration(
                                 node, getCurrentPath(), checker.getContext());
             }
-
-            FlowExpressions.Receiver expr = null;
+            FlowExpressions.Receiver expr;
             try {
-                // TODO: currently, these expressions are parsed at the
-                // declaration (i.e. here) and for every use. this could be
-                // optimized to store the result the first time. (same for
-                // other annotations)
                 expr =
                         FlowExpressionParseUtil.parse(
                                 expression, flowExprContext, getCurrentPath(), false);
-
-                CFAbstractStore<?, ?> exitStore = atypeFactory.getRegularExitStore(node);
-                if (exitStore == null) {
-                    // if there is no regular exitStore, then the method
-                    // cannot reach the regular exit and there is no need to
-                    // check anything
-                } else {
-                    CFAbstractValue<?> value = exitStore.getValue(expr);
-                    AnnotationMirror inferredAnno = null;
-                    if (value != null) {
-                        QualifierHierarchy hierarchy = atypeFactory.getQualifierHierarchy();
-                        Set<AnnotationMirror> annos = value.getAnnotations();
-                        inferredAnno = hierarchy.findAnnotationInSameHierarchy(annos, annotation);
-                    }
-                    if (!checkContract(expr, annotation, inferredAnno, exitStore)) {
-                        checker.report(
-                                Result.failure(
-                                        "contracts.postcondition.not.satisfied", expr.toString()),
-                                node);
+                if (!abstractMethod) {
```

Comment:
It would be nicer if the logic that can't throw the FlowExpressionParseException were outside of this try/catch block, i.e. make the scope of the try/catch as small as possible.